[Bug]: Arctic Define watermark on unkeyed table encounters FieldNotFound Exception

### What happened?

When defining watermark on an Arctic unkey table and executing some dml operations on this table, an exception is encountered that the field does not exist.

### Affects Versions

master

### What engines are you seeing the problem on?

Flink

### How to reproduce

1. create an unkeyed Arctic table
2. define watermark on this unkeyed Arctic table
3. insert into print table select * from this unkeyed Arctic table and then will encounter FieldNotFound Exception

### Relevant log output

```shell
org.apache.flink.table.client.gateway.SqlExecutionException: Invalid SQL statement.
	at org.apache.flink.table.client.gateway.local.LocalExecutor.executeUpdateInternal(LocalExecutor.java:493) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.gateway.local.LocalExecutor.executeUpdate(LocalExecutor.java:428) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.cli.CliClient.callInsert(CliClient.java:690) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.cli.CliClient.callCommand(CliClient.java:327) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at java.util.Optional.ifPresent(Optional.java:159) [?:1.8.0_332]
	at org.apache.flink.table.client.cli.CliClient.open(CliClient.java:214) [flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.SqlClient.openCli(SqlClient.java:144) [flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.SqlClient.start(SqlClient.java:115) [flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.SqlClient.main(SqlClient.java:201) [flink-sql-client_2.12-1.12.7.jar:1.12.7]
Caused by: java.lang.IllegalArgumentException: Field rtime not found in source schema
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.field(ReassignIds.java:77) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.field(ReassignIds.java:28) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.TypeUtil$VisitFieldFuture.get(TypeUtil.java:472) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.Iterators$6.transform(Iterators.java:783) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.TransformedIterator.next(TransformedIterator.java:47) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.Iterators.addAll(Iterators.java:356) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.Lists.newArrayList(Lists.java:143) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.Lists.newArrayList(Lists.java:130) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.struct(ReassignIds.java:55) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.struct(ReassignIds.java:28) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.TypeUtil.visit(TypeUtil.java:505) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.TypeUtil$VisitFuture.get(TypeUtil.java:457) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.schema(ReassignIds.java:40) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.ReassignIds.schema(ReassignIds.java:28) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.TypeUtil.visit(TypeUtil.java:477) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.types.TypeUtil.reassignIds(TypeUtil.java:227) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.flink.FlinkSchemaUtil.convert(FlinkSchemaUtil.java:104) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.flink.source.FlinkSource$Builder.buildFormat(FlinkSource.java:196) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.shaded.org.apache.iceberg.flink.source.FlinkSource$Builder.build(FlinkSource.java:204) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.flink.table.FlinkSource$Builder.buildUnkeyedTableSource(FlinkSource.java:195) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.flink.table.FlinkSource$Builder.build(FlinkSource.java:140) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.flink.table.ArcticFileSource.createDataStream(ArcticFileSource.java:125) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.flink.table.ArcticFileSource.access$000(ArcticFileSource.java:58) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at com.netease.arctic.flink.table.ArcticFileSource$1.produceDataStream(ArcticFileSource.java:192) ~[arctic-flink-runtime-1.12-0.3.2.jar:?]
	at org.apache.flink.table.planner.plan.nodes.common.CommonPhysicalTableSourceScan.createSourceTransformation(CommonPhysicalTableSourceScan.scala:88) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlanInternal(StreamExecTableSourceScan.scala:91) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlanInternal(StreamExecTableSourceScan.scala:44) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecTableSourceScan.translateToPlan(StreamExecTableSourceScan.scala:44) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalc.translateToPlanInternal(StreamExecCalc.scala:54) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalc.translateToPlanInternal(StreamExecCalc.scala:39) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecCalcBase.translateToPlan(StreamExecCalcBase.scala:38) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlanInternal(StreamExecSink.scala:79) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlanInternal(StreamExecSink.scala:43) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan(ExecNode.scala:59) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.exec.ExecNode.translateToPlan$(ExecNode.scala:57) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.plan.nodes.physical.stream.StreamExecSink.translateToPlan(StreamExecSink.scala:43) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.delegation.StreamPlanner.$anonfun$translateToPlan$1(StreamPlanner.scala:66) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.Iterator.foreach(Iterator.scala:937) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.Iterator.foreach$(Iterator.scala:937) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1425) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.IterableLike.foreach(IterableLike.scala:70) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.IterableLike.foreach$(IterableLike.scala:69) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.TraversableLike.map(TraversableLike.scala:233) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.TraversableLike.map$(TraversableLike.scala:226) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at scala.collection.AbstractTraversable.map(Traversable.scala:104) ~[flink-dist_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.delegation.StreamPlanner.translateToPlan(StreamPlanner.scala:65) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.planner.delegation.PlannerBase.translate(PlannerBase.scala:162) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.translate(TableEnvironmentImpl.java:1329) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.translateAndClearBuffer(TableEnvironmentImpl.java:1321) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.api.bridge.java.internal.StreamTableEnvironmentImpl.getPipeline(StreamTableEnvironmentImpl.java:328) ~[flink-table-blink_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.gateway.local.ExecutionContext.lambda$createPipeline$1(ExecutionContext.java:287) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.gateway.local.ExecutionContext.wrapClassLoader(ExecutionContext.java:256) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.gateway.local.ExecutionContext.createPipeline(ExecutionContext.java:282) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	at org.apache.flink.table.client.gateway.local.LocalExecutor.executeUpdateInternal(LocalExecutor.java:490) ~[flink-sql-client_2.12-1.12.7.jar:1.12.7]
	... 8 more
[ERROR] Could not execute SQL statement. Reason:
java.lang.IllegalArgumentException: Field rtime not found in source schema
```


### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -193,7 +193,7 @@ public class FlinkSource {
           .flinkConf(flinkConf)
           .limit(limit)
           .build();
-      return wrapKrb(origin);
+      return wrapKrb(origin).assignTimestampsAndWatermarks(watermarkStrategy);
     }
 
     /**

--- a/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -186,7 +186,7 @@ public class FlinkSource {
     public DataStream<RowData> buildUnkeyedTableSource() {
       DataStream<RowData> origin = org.apache.iceberg.flink.source.FlinkSource.forRowData()
           .env(env)
-          .project(projectedSchema)
+          .project(filterWatermark(projectedSchema))
           .tableLoader(tableLoader)
           .filters(filters)
           .properties(properties)

--- a/flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
+++ b/flink/v1.12/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
@@ -229,6 +229,61 @@ public class TestUnkeyed extends FlinkTestBase {
   }
 
   @Test
+  public void testUnkeyedWatermarkSet() throws Exception {
+    List<Object[]> data = new LinkedList<>();
+
+    data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+    data.add(new Object[]{1000015, "b", LocalDateTime.parse("2022-06-17T10:08:11.0")});
+    data.add(new Object[]{1000011, "c", LocalDateTime.parse("2022-06-18T10:10:11.0")});
+    data.add(new Object[]{1000014, "d", LocalDateTime.parse("2022-06-17T10:11:11.0")});
+    data.add(new Object[]{1000021, "d", LocalDateTime.parse("2022-06-17T16:10:11.0")});
+    data.add(new Object[]{1000007, "e", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+
+    List<ApiExpression> rows = DataUtil.toRows(data);
+
+    Table input = getTableEnv().fromValues(DataTypes.ROW(
+        DataTypes.FIELD("id", DataTypes.INT()),
+        DataTypes.FIELD("name", DataTypes.STRING()),
+        DataTypes.FIELD("ts", DataTypes.TIMESTAMP())
+      ),
+      rows
+    );
+    getTableEnv().createTemporaryView("input", input);
+
+    sql("CREATE CATALOG arcticCatalog WITH %s", toWithClause(props));
+
+    sql("CREATE TABLE IF NOT EXISTS arcticCatalog." + db + "." + TABLE + "(" +
+      " id INT, name STRING, ts TIMESTAMP)" +
+      " WITH (" +
+      " 'location' = '" + tableDir.getAbsolutePath() + "/" + TABLE + "'" +
+      ")");
+
+    sql("create table user_tb (" +
+      "    rtime as cast(ts as timestamp(3))," +
+      "    WATERMARK FOR rtime as rtime" +
+      "  ) LIKE arcticCatalog." + db + "." + TABLE);
+
+    sql("insert into arcticCatalog." + db + "." + TABLE + " select * from input");
+
+    TableResult result = exec("select id, name, ts from user_tb" +
+      "/*+ OPTIONS(" +
+      "'arctic.read.mode'='file'" +
+      ", 'scan.startup.mode'='earliest'" +
+      ")*/" +
+      "");
+
+    Set<Row> actual = new HashSet<>();
+    try (CloseableIterator<Row> iterator = result.collect()) {
+      for (Object[] datum : data) {
+        actual.add(iterator.next());
+      }
+    }
+    Assert.assertEquals(DataUtil.toRowSet(data), actual);
+
+    result.getJobClient().ifPresent(TestUtil::cancelJob);
+  }
+
+  @Test
   public void testSinkBatchRead() throws IOException {
     List<Object[]> data = new LinkedList<>();
     data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});

--- a/flink/v1.14/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.14/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -194,7 +194,7 @@ public class FlinkSource {
           .flinkConf(flinkConf)
           .limit(limit)
           .build();
-      return wrapKrb(origin);
+      return wrapKrb(origin).assignTimestampsAndWatermarks(watermarkStrategy);
     }
 
     /**

--- a/flink/v1.14/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.14/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -187,7 +187,7 @@ public class FlinkSource {
     public DataStream<RowData> buildUnkeyedTableSource() {
       DataStream<RowData> origin = org.apache.iceberg.flink.source.FlinkSource.forRowData()
           .env(env)
-          .project(projectedSchema)
+          .project(filterWatermark(projectedSchema))
           .tableLoader(tableLoader)
           .filters(filters)
           .properties(properties)

--- a/flink/v1.14/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
+++ b/flink/v1.14/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
@@ -183,6 +183,61 @@ public class TestUnkeyed extends FlinkTestBase {
   }
 
   @Test
+  public void testUnkeyedWatermarkSet() throws Exception {
+    List<Object[]> data = new LinkedList<>();
+
+    data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+    data.add(new Object[]{1000015, "b", LocalDateTime.parse("2022-06-17T10:08:11.0")});
+    data.add(new Object[]{1000011, "c", LocalDateTime.parse("2022-06-18T10:10:11.0")});
+    data.add(new Object[]{1000014, "d", LocalDateTime.parse("2022-06-17T10:11:11.0")});
+    data.add(new Object[]{1000021, "d", LocalDateTime.parse("2022-06-17T16:10:11.0")});
+    data.add(new Object[]{1000007, "e", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+
+    List<ApiExpression> rows = DataUtil.toRows(data);
+
+    Table input = getTableEnv().fromValues(DataTypes.ROW(
+        DataTypes.FIELD("id", DataTypes.INT()),
+        DataTypes.FIELD("name", DataTypes.STRING()),
+        DataTypes.FIELD("ts", DataTypes.TIMESTAMP())
+      ),
+      rows
+    );
+    getTableEnv().createTemporaryView("input", input);
+
+    sql("CREATE CATALOG arcticCatalog WITH %s", toWithClause(props));
+
+    sql("CREATE TABLE IF NOT EXISTS arcticCatalog." + db + "." + TABLE + "(" +
+      " id INT, name STRING, ts TIMESTAMP)" +
+      " WITH (" +
+      " 'location' = '" + tableDir.getAbsolutePath() + "/" + TABLE + "'" +
+      ")");
+
+    sql("create table user_tb (" +
+      "    rtime as cast(ts as timestamp(3))," +
+      "    WATERMARK FOR rtime as rtime" +
+      "  ) LIKE arcticCatalog." + db + "." + TABLE);
+
+    sql("insert into arcticCatalog." + db + "." + TABLE + " select * from input");
+
+    TableResult result = exec("select id, name, ts from user_tb" +
+      "/*+ OPTIONS(" +
+      "'arctic.read.mode'='file'" +
+      ", 'scan.startup.mode'='earliest'" +
+      ")*/" +
+      "");
+
+    Set<Row> actual = new HashSet<>();
+    try (CloseableIterator<Row> iterator = result.collect()) {
+      for (Object[] datum : data) {
+        actual.add(iterator.next());
+      }
+    }
+    Assert.assertEquals(DataUtil.toRowSet(data), actual);
+
+    result.getJobClient().ifPresent(TestUtil::cancelJob);
+  }
+
+  @Test
   public void testSinkBatchRead() throws IOException {
     List<Object[]> data = new LinkedList<>();
     data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});

--- a/flink/v1.15/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.15/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -199,7 +199,7 @@ public class FlinkSource {
     public DataStream<RowData> buildUnkeyedTableSource() {
       DataStream<RowData> origin = org.apache.iceberg.flink.source.FlinkSource.forRowData()
           .env(env)
-          .project(projectedSchema)
+          .project(filterWatermark(projectedSchema))
           .tableLoader(tableLoader)
           .filters(filters)
           .properties(properties)

--- a/flink/v1.15/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
+++ b/flink/v1.15/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java
@@ -206,7 +206,7 @@ public class FlinkSource {
           .flinkConf(flinkConf)
           .limit(limit)
           .build();
-      return wrapKrb(origin);
+      return wrapKrb(origin).assignTimestampsAndWatermarks(watermarkStrategy);
     }
 
     /**

--- a/flink/v1.15/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
+++ b/flink/v1.15/flink/src/test/java/com/netease/arctic/flink/table/TestUnkeyed.java
@@ -183,6 +183,61 @@ public class TestUnkeyed extends FlinkTestBase {
   }
 
   @Test
+  public void testUnkeyedWatermarkSet() throws Exception {
+    List<Object[]> data = new LinkedList<>();
+
+    data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+    data.add(new Object[]{1000015, "b", LocalDateTime.parse("2022-06-17T10:08:11.0")});
+    data.add(new Object[]{1000011, "c", LocalDateTime.parse("2022-06-18T10:10:11.0")});
+    data.add(new Object[]{1000014, "d", LocalDateTime.parse("2022-06-17T10:11:11.0")});
+    data.add(new Object[]{1000021, "d", LocalDateTime.parse("2022-06-17T16:10:11.0")});
+    data.add(new Object[]{1000007, "e", LocalDateTime.parse("2022-06-17T10:10:11.0")});
+
+    List<ApiExpression> rows = DataUtil.toRows(data);
+
+    Table input = getTableEnv().fromValues(DataTypes.ROW(
+        DataTypes.FIELD("id", DataTypes.INT()),
+        DataTypes.FIELD("name", DataTypes.STRING()),
+        DataTypes.FIELD("ts", DataTypes.TIMESTAMP())
+      ),
+      rows
+    );
+    getTableEnv().createTemporaryView("input", input);
+
+    sql("CREATE CATALOG arcticCatalog WITH %s", toWithClause(props));
+
+    sql("CREATE TABLE IF NOT EXISTS arcticCatalog." + db + "." + TABLE + "(" +
+      " id INT, name STRING, ts TIMESTAMP)" +
+      " WITH (" +
+      " 'location' = '" + tableDir.getAbsolutePath() + "/" + TABLE + "'" +
+      ")");
+
+    sql("create table user_tb (" +
+      "    rtime as cast(ts as timestamp(3))," +
+      "    WATERMARK FOR rtime as rtime" +
+      "  ) LIKE arcticCatalog." + db + "." + TABLE);
+
+    sql("insert into arcticCatalog." + db + "." + TABLE + " select * from input");
+
+    TableResult result = exec("select id, name, ts from user_tb" +
+      "/*+ OPTIONS(" +
+      "'arctic.read.mode'='file'" +
+      ", 'scan.startup.mode'='earliest'" +
+      ")*/" +
+      "");
+
+    Set<Row> actual = new HashSet<>();
+    try (CloseableIterator<Row> iterator = result.collect()) {
+      for (Object[] datum : data) {
+        actual.add(iterator.next());
+      }
+    }
+    Assert.assertEquals(DataUtil.toRowSet(data), actual);
+
+    result.getJobClient().ifPresent(TestUtil::cancelJob);
+  }
+
+  @Test
   public void testSinkBatchRead() throws IOException {
     List<Object[]> data = new LinkedList<>();
     data.add(new Object[]{1000004, "a", LocalDateTime.parse("2022-06-17T10:10:11.0")});

